text/0000-expr-based-path.md

```diff
@@ -0,0 +1,88 @@
+
+# Motivation
+[motivation]: #motivation
+
+It is sometimes cumbersome to store a result in an intermediate value to extract only a single sub-element from the result. It is also counterintuitive not to use an expression such as `function(arg).something` and can lead to unexpected syntax errors from tremor.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+No new concepts are introduced. The existing concept of a path is extended. As of today, a path could originate from:
+
+- `event` an event
+- `$` the event metadata
+- a local variable
+- a constant
+- the build-in keywords `args`, `group` and `state`
+
+We suggest extending this with four more variants:
+
+- `{...}` record literals
+- `[...]` array literals
+- `struct::keys(a_record)` a function call or a the return for it
+- `(<immutable expr>)` any immutable expression in paratheses
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This is realized by either:
+
+* using the returned reference from the expression as a root for a lookup
+* introducing a new temporary local variable to assign the returned owned variable to and then using a reference to this as a root for the lookup
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The only drawback discovered during this PR is that it introduces a possibility to create less readable scripts where it would have made sense to introduce a named variable for readability, but instead, an expression root was used.
+
+For example the first can be more readable then the second especially if the match statement grows:
+
+```tremor
+let key = match event of
+   case %{present k1} => event.k1
+   case %{present k2} => event.k2
+   case %{present k3} => event.k3
+   case _ => event
+end;
+key.badger
+
+## or:
+
+(match event of
+   case %{present k1} => event.k1
+   case %{present k2} => event.k2
+   case %{present k3} => event.k3
+   case _ => event
+end).badger
+```
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+An alternative would be leaving it as it is, but in many cases, this is counterintuitive. For example, `struct::keys(event)[0]` for "get me the first event key" feels natural while `let keys = struct::keys(event); keys[0]` does not.
+
+# Prior art
+[prior-art]: #prior-art
+
+The current expression syntax is realistically the prior art this lead to, along with many other languages that allow path expressions on functions, expressions, or literals.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+Should we allow **any** expression as a root? Right now, we purposefully exclude mutable expressions to prevent complexity from creeping into scripts. Still, there also is an argument to be made to allow any expression as a root since tremor-script is ultimately an expression-oriented language.
```

Comment:
I would definitely go with all expressions, as least from the grammar point of view. Cases that do not make sense could be discarded in a later compilation step where we care about semantics. Or a warning can be issued about unreadable coding style. (our own builtin clippy! `trippy` FTW)

Reply:
Wrapping this up with a reference to our discussion. We already limit where mutable code can be used. It is not allowed in antying we match on. For example it is not allowed in `[]` so `event[let idx = 6]` is not valid nor is `match let var = "liable" of ... end`. So changing it here would make only sense if we also revisit those decisions and go beyond the scope of this RFC (as we got to ask if we want to even allow this or add more complexity than needed).